packages/plugins/scripts-runner/src/tracker.ts

```diff
@@ -0,0 +1,101 @@
+import * as fs from 'fs-extra';
+import { BlockTransactionObject } from 'web3-eth';
+import Web3 from "web3";
+
+export interface TrackingData {
+  [key: string]: {
+    migrations: string[];
+  };
+}
+
+export interface TrackConfig {
+  scriptName: string;
+  scriptDirectory: string;
+  forceTracking: boolean;
+  migrationsDir?: string;
+}
+
+export interface ScriptsTracker {
+  ensureTrackingFile(hash: string): Promise<void>;
+  track(trackConfig: TrackConfig): Promise<void>;
+  isTracked(scriptName: string): Promise<boolean>;
+  setWeb3(web3: Web3);
+}
+
+export class FileSystemTracker implements ScriptsTracker {
+
+  private _block: BlockTransactionObject | null = null;
+
+  private web3: Web3 | null = null;
+
+  constructor(private trackingFilePath: string, private migrationsDirectory: string) {}
+
+  get block() {
+    return (async () => {
+      if (this._block) {
+        return this._block;
+      }
+      try {
+        this._block = await this.web3?.eth.getBlock(0, true) || null;
```

Comment:
Returning `null` when we don't have web3 sounds very race conditiony. I think either this tracker should get it's own `web3` instance or make sure that we give the web3 instance in its constructor

Reply:
> Returning null when we don't have web3 sounds very race conditiony

So the reason we fall back to `null` here is to satisfy the type checker. Since `web3` could be undefined (hence the elvis operator `?.`) `this._block` could potentially get `undefined` as well. However the type is `BlockTransactionObject | null`. Didn't feel like extending it to `undefined` because we **do** assign a value.

> or make sure that we give the web3 instance in its constructor

^ I believe this isn't so trivial. Passing it as a dependency requires us to pass it inside `ScriptsRunnerPlugin` , which means `ScriptsRunnersPlugin` needs to get it injected as well (since we don' want constructors to be asyncronous).

This unfortunately isn't possible with Embark's current plugin system as similarly discussed [here](https://github.com/embarklabs/embark/pull/2253#discussion_r377564196).

I'll go with duplicating the web3 getter instead.